paasta_tools/autoscaling_lib.py

```diff
@@ -514,41 +547,165 @@ def spotfleet_scaler(resource, error):
 
     if new_capacity != current_capacity:
         print "Scaling SFR %s from %d to %d!" % (resource['id'], current_capacity, new_capacity)
-        scale_aws_spot_fleet_request(sfr_id=resource['id'], target_capacity=new_capacity, ec2_client=ec2_client)
+        return new_capacity - current_capacity, new_capacity
     else:
-        print "No need to scale. new_capacity (%d) matches current capacity (%d)" % (new_capacity, current_capacity)
+        return 0
+
 
+def sort_slaves_to_kill(mesos_state, pool='default'):
+    """Pick the best slaves to kill. This returns a list of slaves
+    after sorting in preference of which slaves we kill first.
+    It sorts first by number of chronos tasks, then by total number of tasks
+
+    :param mesos_state: mesos_state dict
+    :param pool: pool of slaves to consider
+    :returns: list of slaves"""
+    slaves = get_mesos_task_count_by_slave(mesos_state, pool=pool)
+    if slaves:
+        slaves_by_task_count = [slave['slave'] for slave in sorted(slaves.values(),
+                                                                   key=lambda x: (x['chronos_count'], x['count']))]
+        return slaves_by_task_count
+    else:
+        return []
 
-def scale_aws_spot_fleet_request(sfr_id, target_capacity, ec2_client):
-    return ec2_client.modify_spot_fleet_request(SpotFleetRequestId=sfr_id, TargetCapacity=target_capacity)
-    #########
-    # TODO: determine which boxes to kill first
-    # TODO: drain boxes/maintainance mode before killing
-    #########
-    # http://boto3.readthedocs.org/en/latest/reference/services/ec2.html#EC2.Client.modify_spot_fleet_request
-    # describes how to use boto to scale down spot fleet requests without
-    # killing instance using the noTermination
-    # ExcessCapacityTerminationPolicy
-    #########
+
+def wait_and_terminate(instances_to_kill, dry_run):
+    """Currently kills a list of instances, will wait for draining to complete soon
+
+    :param instances_to_kill: dict of instances to kill
+    :param dry_run: Don't drain or make changes to spot fleet if True"""
+    ec2_client = boto3.client('ec2')
+    try:
+        # This loop should always finish because the maintenance window should trigger is_ready_to_kill
+        # being true. Just in case though we set a timeout and terminate anyway
+        with Timeout(seconds=CLUSTER_DRAIN_TIMEOUT + 300):
+            while instances_to_kill:
+                for slave_pid, slave_data in instances_to_kill.items():
+                    instance_id = slave_data['instance_id']
+                    if not instance_id:
+                        # if we failed to find instance id then skip
+                        # shouldn't happen but just incase
+                        continue
+                    is_ready_to_kill = True
```

Comment:
Agreed. So my initial plan is for the call to paasta maintenance to check the slave is running no tasks OR it has reached the maintenance window (which means it failed to drain in a timely fasion).

I'm thinking I might handle the `is_not_in_mesos` earlier on because by this point we're only looking at slaves that we found.